pulser-simulation/pulser_simulation/operators.py

```diff
@@ -0,0 +1,278 @@
+            computational basis in which the operators are defined.
+        operation: A string: either "I" or "sigma_{a}{b}" with a, b in the
+            computational basis.
+
+    Returns:
+        The desired operator as a qutip.Qobj object, identity if "I",
+        ``sigma_{ab} = a * b^\dagger`` if "sigma_ab".
+    """
+    return build_operator(
+        sampled_seq,
+        {0: [0.0, 0.0]},
+        [(operation, [0])],
+    )
+
+
+def build_1qubit_operator(
+    sampled_seq: SequenceSamples,
+    operations: list[tuple[float, list[str | qutip.Qobj]]],
+    operators: Mapping[str, qutip.Qobj] | None = None,
+) -> qutip.Qobj:
+    r"""Creates a 1 qubit operator summing projectors and qutip objects.
+
+    Takes as argument a list of operations to apply on one qubit. Returns
+    the sum of these operations. The elements in operations can be a
+    ``qutip.Qobj`` or a string key. If ``operators`` is undefined, this string
+    can be "I" or "sigma_{a}{b}" with a, b elements of the computational basis
+    (sampled_seq.eigenbasis with "x" if with_leakage is True) and
+    ``sigma_{ab} = a * b^\dagger``.
+
+    Examples:
+        If we have a sampled sequence with only a Rydberg Channel, by default
+        the available operators are "I" and "sigma_ab" with a, b in ["r", "g"].
+        The operator ``sigma_gg - sigma_rr`` can be written as::
+
+            ```python
+                build_1qubit_operator(
+                    sampled_seq,
+                    [(1.0, "sigma_gg"), (-1.0, "sigma_rr")],
+                    with_leakage=True,
+                )
+            ```
+
+        We can also define our custom set of operators. For instance, by
+        setting `operators = {"X":qutip.sigmax(), "Z":qutip.sigmaz()}` you can
+        make combinations of these operators such as X-Z::
+
+        ```python
+            build_1qubit_operator(
+                sampled_seq,
+                [(1.0, "X"), (-1.0, "Z")],
+                operators=operators,
+            )
+        ```
+
+    Args:
+        sampled_seq: The SequenceSamples to consider. Provides the
+            computational basis in which the operators are defined.
+        operations: List of tuples `(operator, qubits)`.
+        operators: A dict of operators and their labels. If None, it is
+            composed of all the projectors on the computational basis
+            (with error state if with_leakage is True) and "I".
+
+
+    Returns:
+        The final operator as a qutip.Qobj object.
+    """
+    if operators is None:
+        operators = default_operators(sampled_seq)
+    return sum(
+        [
+            operation[0]
+            * build_operator(
+                sampled_seq,
+                {0: [0.0, 0.0]},
+                [(operation[1], [0])],
+                operators,
+            )
+            for operation in operations
+        ]
+    )
+
+
+def build_operator(
+    sampled_seq: SequenceSamples,
```

Comment:
You feed the sequence samples to get the used basis, for which I see you've added a method to `SequenceSamples`. What is the reason you've added it here, and not to `Sequence`. Now you force the user to discretize the sequence  whenever they want to define an operator, while in principle, that should be an operation the backend should do as part of its internal workings, no?

Reply:
Alternatively, if all you care about from the `SequenceSamples` is to obtain a tuple like `('r','g')` from it, why not ask the user to feed in the basis like that directly. In that case, the user can get it from either `Sequence` or `SequenceSamples` since I see methods in both of them for getting the basis. It has the added benefit of making it more explicit what information is actually used by these functions, and also that you can use it on an equal footing with `Backend`s which take the `Sequence` and `Simulator`s which take the `SequenceSamples`, since this API then no longer requires either of those.